### ISSUE TITLE
[BUG] Missing ops in profiler

### ISSUE THREAD

**Issue by rwightman** [microsoft/DeepSpeed#1467]
**Describe the bug**
I recently discovered the DeepSpeed profiler and like that it's super easy to use and the layer / model view is quite intuitive and informative. However, digging in deeper I'm seeing some issues re several transformer models. Comparing familiar CNNs the flop counts seem accurate and very closer to past analysis. For transformers with matmuls ops in custom modules, those matmuls look like they aren't recorded (and likely all other ops that are tensor ops (torch.matmul, @, *, torch.mul, etc). I assume the comment about nn.functional being captured, Megatron working implied tensor ops as well, but perhaps not? 

See example for an attn module where matmuls appear to be missed: https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py#L195 ... also modules like https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/halo_attn.py#L195

I've done some comparisons with fvcore's aten op approach (via tracing) and it's catching the matmuls but is listing quite a few less significant aten ops as unsupported.

**Expected behavior**
It's common to have explicit torch.matmul, other tensor ops, that can have fairly significant flops for models like transformers. I'd be nice if these to be patched somehow and included... although guessing that could be challenging without profiling at a lower level or tracing the model.

**Docker context**
Running in NGC 21.09 container w/ DeepSpeed 0.5.4 via pip

**Comment by cli99** [comment]
> I spent a little time looking at this, and it appears that that mechanism used for patching the nn.functional methods could also be used for tensor methods...
> 
> I added this as a quick test and was able to recover most of the missing flops relative to fvcore, actually a bit more since some of the pointwise add, mul, and gelu are missed by them
> 
> ```
>     torch.matmul = wrapFunc(torch.matmul, _matmul_fc)
>     torch.Tensor.__matmul__ = wrapFunc(torch.Tensor.__matmul__, _matmul_fc)
> 
>     torch.mul = wrapFunc(torch.mul, _pointw_fc)
>     torch.Tensor.__mul__ = wrapFunc(torch.Tensor.__mul__, _pointw_fc)
>     torch.Tensor.__imul__ = wrapFunc(torch.Tensor.__imul__, _pointw_fc)
>     
>     torch.add = wrapFunc(torch.add, _pointw_fc)
>     torch.Tensor.__add__ = wrapFunc(torch.Tensor.__add__, _pointw_fc)
>     torch.Tensor.__iadd__ = wrapFunc(torch.Tensor.__iadd__, _pointw_fc)
> ```
> 
> Also a number of fairly common nn.functional is missing
> 
> ```
>     F.silu = wrapFunc(F.silu, ...)
>     F.gelu = wrapFunc(F.gelu, ...)
>     F.layer_norm = ...
>     F.instance_norm = ...
> ```
> 
> EDIT: I have no idea if my approach here was correct in all cases, but it did appear to work
Hi rwightman, thanks for reporting the issue. Your approach of adding flops counting to tensor methods is correct. Can you please submit of PR of your change? Also if you have any flops calculation for the common nn functionals that are missing, please contribute as well. Thanks.

**Comment by cli99** [comment]
> One other related comment. This profiler is counting some pointwise (activations, softmax) and reduction/pooling ops that are often ignored (for GMACs focused counters).
> 
> Would it make sense to track GMACs and FLOPs separately? ie include pointwis, reduction, pooling ops (and broaden to include others like add, mul, sub, mean, etc) in FLOPs but not add them to GMACs?

FLOPs is computed as 2* GMACs / duration of the module. Can you elaborate more on why to track GMACs and FLOPs separately? Thanks.

**Comment by rwightman** [comment]
@cli99 FLOPs as in floating point operations (plural) not per second. Generally what most frameworks report as FLOPs is actually GMACs, and 2* is an approximation.The operations that relate directly to multiply-accumulates are counted as GMACs (ie matmuls, convs, einsum, etc). Operations such as pointwise adds, multiplies, activations, are not typically included as there is no sensible translation between them in terms of GMACs and thus the approximate FLOPs = 2*GMACS doesn't make sense.

So the point was one could sum ALL operations for a complete view of FLOPs (including non MAC ops), and have that in addition to the GMACs calculation.

This is all separate from the fact that the profiler her isn't tracking matmuls and einsum which are core operations for GMACs, and it is including some ops that are typically not included.

**Comment by cli99** [comment]
> Also, conv flop ops don't actually support 1d and 3d properly, they are hardcoded to 2d based on
> 
> ```
>     paddings = padding if type(padding) is tuple else (padding, padding)
>     strides = stride if type(stride) is tuple else (stride, stride)
>     dilations = dilation if type(dilation) is tuple else (dilation, dilation)
> 
>     output_dims = [0, 0]
>     output_dims[0] = (input_dims[0] + 2 * paddings[0] -
>                       (dilations[0] * (kernel_dims[0] - 1) + 1)) // strides[0] + 1
>     output_dims[1] = (input_dims[1] + 2 * paddings[1] -
>                       (dilations[1] * (kernel_dims[1] - 1) + 1)) // strides[1] + 1
> ```

Thanks for pointing this out. This PR should fix it. https://github.com/microsoft/DeepSpeed/pull/1518

**Comment by jeffra** [comment]
Please re-open if #1518 didn't fix the issue or if there are further issues. Thank you @rwightman :)

**Comment by rwightman** [comment]
@jeffra #1518 is only part of the issue and secondary to the main problem that man important matmul type operations that occur outside of nn.functional or nn.Module based layers are missed
* torch.matmul, torch.mm, torch.baddbmm, torch.bmm, etc are missed
* Tensor.matmul, `x @ y`, etc are missed
* some activations are missing (relu and elu present but no sigmoid, silu, and gelu as examples)
* batch norm is included but other norms are missing

And then there was also the mentioned consistency with regards to gmacs vs flop counting. You include activations (or rather some of them) but other implementations that report FLOPs as GMACs (like you do) tend to only include the operations that can be directly translated to fused multiply-add ops (like matmuls, etc).

**Comment by jeffra** [comment]
Thanks for clarifying the open issues, sorry for premature closing :) Just discussed with @cli99, she's going to look at addressing these pending issues.

**Comment by cli99** [comment]
Hi @rwightman, I created this PR to address the issues you brought up. Please take a look and any feedback is welcomed. https://github.com/microsoft/DeepSpeed/pull/1591

**Comment by B06901052** [comment]
Hello, it seems there is still a line hardcoded to 2d at [here](https://github.com/microsoft/DeepSpeed/blob/89b8c8872bad3f3e11a846e97d4cb1fb796e554d/deepspeed/profiling/flops_profiler/profiler.py#L537), which may be wrong for 1d and 3d ([-2:] makes len(kernel_dims)=2).
```python
576    kernel_dims = list(weight.shape[-2:])
```
